### PR TITLE
sql: move NullableArgs function property to overload level

### PR DESCRIPTION
**Pull request by chengxiong-ruan**
Currently we only have builtins, and that all overloads of a
same function name share function properties. However, we're
going to support user defined functions whose properties can
vary as how users define them. So we need to move any property
that's relevant to UDF to overload level. Currently, `NullableArgs`
is the only one matters.

Release note: None.

### CODE REVIEW

**Comment by chengxiong-ruan** [comment]
CI failures seems not relevant

File: pkg/sql/sem/tree/type_check.go [hunk 1]
```diff
@@ -1071,17 +1071,27 @@ func (expr *FuncExpr) TypeCheck(
...
 	}
 
+	var nullableArgFns []overloadImpl
+	var unNullableArgFns []overloadImpl
+	for _, f := range fns {
+		if f.(*Overload).NullableArgs {
+			nullableArgFns = append(nullableArgFns, f)
+			continue
+		}
+		unNullableArgFns = append(unNullableArgFns, f)
```
Comment (ajwerner) [inline pkg/sql/sem/tree/type_check.go:1081]: nit: this seems like a nice place for an `else`

File: pkg/sql/sem/builtins/aggregate_builtins.go [hunk 2]
```diff
@@ -690,29 +651,46 @@ func makePrivate(b builtinDefinition) builtinDefinition {
...
 		tree.FixedReturnType(ret),
 		f,
 		info,
 		volatility,
+		nullableArgs,
 	)
 }
 
+func makeDefaultAggOverloadWithReturnType(
+	in []*types.T, retType tree.ReturnTyper, f eval.AggregateOverload, info string,
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/sql/sem/builtins/aggregate_builtins.go:682]: Feel free to not act on this one.

**Review by ajwerner — APPROVED** [review]
LGTM though please do consider renaming `unNullable`

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83732)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, and @rafiss)

___
*[`pkg/sql/sem/builtins/aggregate_builtins.go` line 654 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83732#-N6J7n_h6DMI_veVnxQH:-N6J7n_h6DMI_veVnxQI:bmzehn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd25e760be104b51d7ebfab88881d34fede7cd99/pkg/sql/sem/builtins/aggregate_builtins.go#L654)):*
> ```Go
> 	return b
> }
> func makeDefaultAggOverload(
> ```

should we name this `makeImmutableAggOverload` (and ditto below)?

**Review by chengxiong-ruan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83732)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, and @rafiss)

___
*[`pkg/sql/sem/builtins/aggregate_builtins.go` line 654 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83732#-N6J7n_h6DMI_veVnxQH:-N6JrEve6Z0NLW1ON81o:b-h8z1h7) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd25e760be104b51d7ebfab88881d34fede7cd99/pkg/sql/sem/builtins/aggregate_builtins.go#L654)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should we name this `makeImmutableAggOverload` (and ditto below)?
</blockquote></details>

yes, good point, changed to that.
___
*[`pkg/sql/sem/tree/type_check.go` line 1075 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83732#-N5vwA8tG1QcLV4sqTMQ-r1-1075:-N6JrICP4brOQ9EMTMue:be0igfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd25e760be104b51d7ebfab88881d34fede7cd99/pkg/sql/sem/tree/type_check.go#L1075)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

What about `notNullableArgsFn`?
</blockquote></details>

done, sorry for the bad naming
___
*[`pkg/sql/sem/tree/type_check.go` line 1081 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83732#-N5vwA8tG1QcLV4sqTMQ-r1-1081:-N6JrMH818OAygs2bJwg:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd25e760be104b51d7ebfab88881d34fede7cd99/pkg/sql/sem/tree/type_check.go#L1081)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: this seems like a nice place for an `else`
</blockquote></details>

done

**Comment by chengxiong-ruan** [comment]
TFTR!
bors r+

**Pull request merged**